### ISSUE TITLE
UninitializedFieldError when using scala compiler option "-Xcheckinit"

### ISSUE THREAD

**Issue by Juraj-Masiar** [jOOQ/jOOQ#5489]
I'm using jOOQ in Play Framework 2.5.4 and after I upgraded and regenerated jOOQ from 2.7.4 to 2.8.4 I'm getting this runtime exception:

```
2016-08-10 12:44:32 - [error] - a.a.ActorSystemImpl - Uncaught fatal error from thread [application-contexts.database-6] shutting down ActorSystem [application]
java.lang.ExceptionInInitializerError: null
        at jooq_gen.Public.<init>(Public.scala:74)
        at jooq_gen.Public$.<init>(Public.scala:61)
        at jooq_gen.Public$.<clinit>(Public.scala)
        at jooq_gen.tables.Person.<init>(Person.scala:53)
        at jooq_gen.tables.Person.<init>(Person.scala:186)
Caused by: scala.UninitializedFieldError: Uninitialized field: Public.scala: 61
        at jooq_gen.Public$.PUBLIC(Public.scala:61)
        at jooq_gen.DefaultCatalog.<init>(DefaultCatalog.scala:42)
        at jooq_gen.DefaultCatalog$.<init>(DefaultCatalog.scala:24)
        at jooq_gen.DefaultCatalog$.<clinit>(DefaultCatalog.scala)
        at jooq_gen.Public.<init>(Public.scala:74)
```

After some googling I find out it's connected with scala compile option "-Xcheckinit" that I was using. Without that everything works fine but still sounds like a bug.

**Comment by lukaseder** [comment]
Thank you very much for reporting.

Are you sure this is a jOOQ bug? I would suspect that this means there is something wrong in the bytecode that's produced by the Scala compiler... What Scala version are you using?

**Comment by icl7126** [comment]
I'm using current 2.11.8 version.  
And these are the compiler options from my `build.sbt`:
```
scalacOptions in ThisBuild ++= Seq(
  "-target:jvm-1.8",
  "-encoding", "UTF-8",
  "-deprecation", // warning and location for usages of deprecated APIs
  "-feature", // warning and location for usages of features that should be imported explicitly
  "-unchecked", // additional warnings where generated code depends on assumptions
  "-Xlint", // recommended additional warnings
//  "-Xcheckinit", // runtime error when a val is not initialized due to trait hierarchies (instead of NPE somewhere else)
  "-Ywarn-adapted-args", // Warn if an argument list is modified to match the receiver
  "-Ywarn-value-discard", // Warn when non-Unit expression results are unused
  "-Ywarn-inaccessible",
  "-Ywarn-dead-code"
)
```
Sorry I have no idea where the bug is ..., I just know it appeared right after the jOOQ update.  
Anyway I can live without `Xcheckinit` so for me the problem is solved. And thanks to this issue report anybody can google it much easier than I had to. So unless more people complain, there is no hurry... maybe the problem get solved 'by itself' with upcoming version of scala 2.12.

**Comment by lukaseder** [comment]
Thanks for the additional feedback. I still think we should get this solved somehow, as it might be annoying to quite a few users. The jOOQ update changed the way some jOOQ generated artefacts are initialised. Stuff still compiles fine with the Scala compiler, which is why I suspect it's not jOOQ's "fault", here (code that compiles should also run).

In the past, I've been talking with @lrytz, a friend of mine who works on the Scala compiler for Lightbend about issues like these. It appears that getting bytecode validation right for the verifier has always been an interesting topic for the Scala compiler. Perhaps, @lrytz has some idea?

**Comment by lrytz** [comment]
The crash is not in bytecode validation, it's a check generated by the Scala compiler (under the `-Xcheckinit` flag). I don't have a good guess to offer.. A reproducible example would help of course, ideally as small as possible.

**Comment by lukaseder** [comment]
We couldn't reproduce this when it was reported.

**Comment by lukaseder** [comment]
I can reproduce this now, see discussion here: https://github.com/jOOQ/jOOQ/issues/8139

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/8139 is fixed in 3.14.0 and 3.13.6
